casper/src/main/scala/coop/rchain/casper/genesis/Genesis.scala

```diff
@@ -103,6 +94,62 @@ object Genesis {
                      )
     } yield genesisBlock
 
+  private def getVaults[F[_]: Sync: Log: RaiseIOError](
```

Comment:
We should not be using sync here. raiseError is available on `FunctorRaise`

Reply:
We should probably also have some custom exception types (at least one - ConfigurationError?) for the error cases we have here

Reply:
Hmm; do you mean I should use `raiseIOErrorThroughSync`?

Reply:
I think @ArturGajowy means that since you have `RaiseIOError`, your can use `RaiseIOError[F].raise(...)` instead of `Sync[F].raiseError(...)`

Reply:
How should I treat

```scala
      lines <- SourceIO
                .open(vaultsPath)
                .use(_.getLines)
                .adaptError {
                  case ex: Throwable =>
                    new RuntimeException(
                      s"Failed to read ${vaultsPath.toAbsolutePath} for reason: ${ex.getMessage}"
                    )
                }
```?

Reply:
Also

```scala
  private def parseLine[F[_]: Sync: Log: RaiseIOError](line: String): F[Vault] =
    Sync[F].fromEither(
      fromLine(line)
        .leftMap(errMsg => new RuntimeException(s"Error in parsing vaults file: $errMsg"))
    )
```

Reply:
@itegulov Wouldn't I still need `Sync` for `ifM`?

Reply:
Yeah, seems like you need at least `ApplicativeError` here for what you are trying to do. `FunctorRaise` is not enough. Maybe try to change `RaiseIOError`'s type to `ApplicativeError`? If it is not as simple/does not make sense I would just stick with `Sync` here...

Regarding `ifM`: you only need `Monad` for it which far less powerful than `Sync`.

Reply:
The Sync is only needed, because we don't yet have a RaiseCasperError type that could inject errors from downstream modules into its own error type. There's not even the error type, as signified by the `new RuntimeException`. It seems to me that we should discuss an approach to error handling at some time in the future, and do small steps refactoring towards it while we progress with PRs.